File: Bots/DotNet/Consumers/CodeFirst/WaterfallHostBot/Skills/WaterfallSkill.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System;
using System.Collections.Generic;
using Microsoft.Bot.Schema;

namespace Microsoft.BotFrameworkFunctionalTests.WaterfallHostBot.Skills
{
    public class WaterfallSkill : SkillDefinition
    {
        private const string SkillActionCards = "Cards";
        private const string SkillActionProactive = "Proactive";
        private const string SkillActionAuth = "Auth";
        private const string SkillActionMessageWithAttachment = "MessageWithAttachment";
        private const string SkillActionSso = "Sso";
        private const string SkillActionFileUpload = "FileUpload";
        private const string SkillActionCallEchoSkill = "Echo";

        public override IList<string> GetActions()
        {
            return new List<string>
            {
                SkillActionCards,
                SkillActionProactive,
                SkillActionAuth,
                SkillActionMessageWithAttachment,
                SkillActionSso,
                SkillActionFileUpload,
                SkillActionCallEchoSkill
            };
        }

        public override Activity CreateBeginActivity(string actionId)
        {
            Activity activity;

            // Send an event activity to the skill with "Cards" in the name.
            if (actionId.Equals(SkillActionCards, StringComparison.CurrentCultureIgnoreCase))
            {
                activity = (Activity)Activity.CreateEventActivity();
                activity.Name = SkillActionCards;
                return activity;
            }

            // Send an event activity to the skill with "Proactive" in the name.
            if (actionId.Equals(SkillActionProactive, StringComparison.CurrentCultureIgnoreCase))
            {
                activity = (Activity)Activity.CreateEventActivity();
                activity.Name = SkillActionProactive;
                return activity;
            }

            // Send an event activity to the skill with "Auth" in the name.
            if (actionId.Equals(SkillActionAuth, StringComparison.CurrentCultureIgnoreCase))
            {
                activity = (Activity)Activity.CreateEventActivity();
                activity.Name = SkillActionAuth;
                return activity;
            }

            // Send an event activity to the skill with "Attachment" in the name.
            if (actionId.Equals(SkillActionMessageWithAttachment, StringComparison.CurrentCultureIgnoreCase))
            {
                activity = (Activity)Activity.CreateEventActivity();
                activity.Name = SkillActionMessageWithAttachment;
                return activity;
            }

            // Send an event activity to the skill with "Sso" in the name.
            if (actionId.Equals(SkillActionSso, StringComparison.CurrentCultureIgnoreCase))
            {
                activity = (Activity)Activity.CreateEventActivity();
                activity.Name = SkillActionSso;
                return activity;
            }

            // Send an event activity to the skill with "FileUpload" in the name.
            if (actionId.Equals(SkillActionFileUpload, StringComparison.CurrentCultureIgnoreCase))
            {
                activity = (Activity)Activity.CreateEventActivity();
                activity.Name = SkillActionFileUpload;
                return activity;
            }

            // Send an event activity to the skill with "Echo" in the name.
            if (actionId.Equals(SkillActionCallEchoSkill, StringComparison.CurrentCultureIgnoreCase))
            {
                activity = (Activity)Activity.CreateEventActivity();
                activity.Name = SkillActionCallEchoSkill;
                return activity;
            }

            throw new InvalidOperationException($"Unable to create begin activity for \"{actionId}\".");
        }
    }
}

File: Bots/Python/Consumers/CodeFirst/WaterfallHostBot/skills/waterfall_skill.py
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License.

from botbuilder.schema import Activity
from skills.skill_definition import SkillDefinition

SKILL_ACTION_CARDS = "Cards"
SKILL_ACTION_PROACTIVE = "Proactive"
SKILL_ACTION_AUTH = "Auth"
SKILL_ACTION_MESSAGE_WITH_ATTACHMENT = "MessageWithAttachment"
SKILL_ACTION_SSO = "Sso"
SKILL_ACTION_FILE_UPLOAD = "FileUpload"
SKILL_ACTION_CALL_ECHO_SKILL = "Echo"


class WaterfallSkill(SkillDefinition):
    def get_actions(self):
        return [
            SKILL_ACTION_CARDS,
            SKILL_ACTION_PROACTIVE,
            SKILL_ACTION_AUTH,
            SKILL_ACTION_MESSAGE_WITH_ATTACHMENT,
            SKILL_ACTION_SSO,
            SKILL_ACTION_FILE_UPLOAD,
            SKILL_ACTION_CALL_ECHO_SKILL,
        ]

    def create_begin_activity(self, action_id: str):
        activity = Activity.create_event_activity()

        if action_id == SKILL_ACTION_CARDS:
            activity.name = SKILL_ACTION_CARDS

        elif action_id == SKILL_ACTION_PROACTIVE:
            activity.name = SKILL_ACTION_PROACTIVE

        elif action_id == SKILL_ACTION_AUTH:
            activity.name = SKILL_ACTION_AUTH

        elif action_id == SKILL_ACTION_MESSAGE_WITH_ATTACHMENT:
            activity.name = SKILL_ACTION_MESSAGE_WITH_ATTACHMENT

        elif action_id == SKILL_ACTION_SSO:
            activity.name = SKILL_ACTION_SSO

        elif action_id == SKILL_ACTION_FILE_UPLOAD:
            activity.name = SKILL_ACTION_FILE_UPLOAD

        elif action_id == SKILL_ACTION_CALL_ECHO_SKILL:
            activity.name = SKILL_ACTION_CALL_ECHO_SKILL

        else:
            raise Exception(
                f'[WaterfallSkill]: Unable to create begin activity for "{ action_id }".'
            )

        return activity
